jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java

```diff
@@ -730,7 +730,12 @@ public int fill(ByteBuffer buffer) throws IOException
 
                                 case BUFFER_UNDERFLOW:
                                     if (netFilled > 0)
-                                        continue; // try filling some more
+                                    {
+                                        if (BufferUtil.space(_encryptedInput) > 0)
+                                            continue; // try filling some more
+                                        BufferUtil.clear(_encryptedInput);
+                                        throw new SSLHandshakeException("Encrypted buffer max length exceeded");
+                                    }
```

Comment:
I think the following is better/safer:

```suggestion
                                        continue;
                                    if (netFilled == 0 && BufferUtil.space(_encryptedInput) == 0)
                                    {
                                        BufferUtil.clear(_encryptedInput);
                                        throw new SSLHandshakeException("Encrypted buffer max length exceeded");
                                    }
```